packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2241,75 +2329,48 @@ export class KeyringController extends BaseController<
    * keyrings contained in it, using a password or an encryption key with salt.
    *
-   * @param password - The keyring controller password.
-   * @param encryptionKey - An exported key string to unlock keyrings with.
-   * @param encryptionSalt - The salt used to encrypt the vault.
+   * @param credentials - The credentials to unlock the keyrings.
    * @returns A promise resolving to the deserialized keyrings array.
    */
   async #unlockKeyrings(
-    password: string | undefined,
-    encryptionKey?: string,
-    encryptionSalt?: string,
+    credentials:
+      | {
+          password: string;
+        }
+      | {
+          exportedEncryptionKey: string;
+        },
   ): Promise<{
     keyrings: { keyring: EthKeyring; metadata: KeyringMetadata }[];
     newMetadata: boolean;
   }> {
     return this.#withVaultLock(async () => {
-      const encryptedVault = this.state.vault;
-      if (!encryptedVault) {
+      if (!this.state.vault) {
         throw new Error(KeyringControllerError.VaultError);
       }
+      const parsedEncryptedVault = JSON.parse(this.state.vault);
 
-      let vault;
-      const updatedState: Partial<KeyringControllerState> = {};
-
-      if (this.#cacheEncryptionKey) {
-        assertIsExportableKeyEncryptor(this.#encryptor);
-
-        if (password) {
-          const result = await this.#encryptor.decryptWithDetail(
-            password,
-            encryptedVault,
-          );
-          vault = result.vault;
-          this.#password = password;
-
-          updatedState.encryptionKey = result.exportedKeyString;
-          updatedState.encryptionSalt = result.salt;
-        } else {
-          const parsedEncryptedVault = JSON.parse(encryptedVault);
-
-          if (encryptionSalt !== parsedEncryptedVault.salt) {
-            throw new Error(KeyringControllerError.ExpiredCredentials);
-          }
-
-          if (typeof encryptionKey !== 'string') {
-            throw new TypeError(KeyringControllerError.WrongPasswordType);
-          }
-
-          const key = await this.#encryptor.importKey(encryptionKey);
-          vault = await this.#encryptor.decryptWithKey(
-            key,
-            parsedEncryptedVault,
-          );
-
-          // This call is required on the first call because encryptionKey
-          // is not yet inside the memStore
-          updatedState.encryptionKey = encryptionKey;
-          // we can safely assume that encryptionSalt is defined here
-          // because we compare it with the salt from the vault
-          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-          updatedState.encryptionSalt = encryptionSalt!;
-        }
+      if ('password' in credentials) {
+        await this.#deriveEncryptionKey(credentials.password);
       } else {
-        if (typeof password !== 'string') {
-          throw new TypeError(KeyringControllerError.WrongPasswordType);
-        }
+        const { exportedEncryptionKey } = credentials;
```

Comment:
Is it ok to assume that it will be defined in the runtime? Maybe we could use something like:

```typescript
if ('password' in credentials) {
  // ...
} else if ('exportedEncryptionKey' in credentials) {
  // ...
} else {
  // Handle error
}
```

Reply:
Note also that if we include the check in the if/else block we'll still need to make TypeScript happy with the additional check I quoted in the previous comment, as `this.#encryptionKey` may still be undefined

Reply:
Though, we could make this prettier by moving it to another internal method that handles the credentials, moving complexity out of this method

Reply:
> I find this a bit confusing because we validate the `password` directly, but delegate the validation of the `exportedEncryptionKey ` 

Where do you see the validation of the `password`? I believe `#unlockKeyrings` does not validate the password type, it's delegated to the `#deriveEncryptionKey` method

Reply:
Sorry, I meant the `if ('password' in credentials) {` check.